third_party/move/move-vm/runtime/src/loader/modules.rs

```diff
@@ -29,36 +29,68 @@ use move_core_types::{
 use move_vm_types::loaded_data::runtime_types::{
     StructIdentifier, StructNameIndex, StructType, Type,
 };
+use parking_lot::RwLock;
 use std::{
     collections::{BTreeMap, HashMap},
     fmt::Debug,
     sync::Arc,
 };
 
-// A ModuleCache is the core structure in the Loader.
-// It holds all Modules, Types and Functions loaded.
-// Types and Functions are pushed globally to the ModuleCache.
-// All accesses to the ModuleCache are under lock (exclusive).
-#[derive(Clone)]
-pub(crate) struct ModuleCache {
-    pub(crate) modules: BinaryCache<ModuleId, Module>,
+/// This trait provides an additional api for the Session to decide where the resolved modules should be stored.
+///
+/// The default api will store the modules inside MoveVM structure but the caller can also choose to store it
+/// elsewhere as long as it implements this `ModuleStorage` trait. Doing so would allow the the caller, i.e: the
+/// adapter layer, to freely decide when to drop or persist the cache as well as determining its own eviction policy.
+pub trait ModuleStorage {
+    fn store_module(&self, module_id: &ModuleId, binary: Module) -> Arc<Module>;
+    fn fetch_module(&self, module_id: &ModuleId) -> Option<Arc<Module>>;
 }
 
+pub(crate) struct ModuleCache(RwLock<BinaryCache<ModuleId, Module>>);
+
 impl ModuleCache {
-    pub(crate) fn new() -> Self {
-        Self {
-            modules: BinaryCache::new(),
-        }
+    pub fn new() -> Self {
+        ModuleCache(RwLock::new(BinaryCache::new()))
+    }
+
+    pub fn flush(&self) {
+        *self.0.write() = BinaryCache::new();
+    }
+}
+
+impl Clone for ModuleCache {
+    fn clone(&self) -> Self {
+        ModuleCache(RwLock::new(self.0.read().clone()))
+    }
+}
+
+impl ModuleStorage for ModuleCache {
+    fn store_module(&self, module_id: &ModuleId, binary: Module) -> Arc<Module> {
+        self.0.write().insert(module_id.clone(), binary).clone()
+    }
+
+    fn fetch_module(&self, module_id: &ModuleId) -> Option<Arc<Module>> {
+        self.0.read().get(module_id).map(Arc::clone)
+    }
+}
+
+pub(crate) struct ModuleStorageAdapter {
+    modules: Arc<dyn ModuleStorage>,
```

Comment:
If ModuleStorage is provided as a remote Arc (which we intend to switch to), we will likely require another local cache for modules, so
(a) things can be cached locally in a single-thread context without having to resolve remotely and paying concurrency overhead (and as a side benefit already providing some additional consistency at this level). 
(b) can be useful for staging when publishing new modules / init_module etc.

If we agree on this direction, then question is whether we can / should implement this already, and what's the best way: e.g. if Arc is just wrapping the local storage, for compatibility, does it make sense to have an extra cache in front, which is the same thing but just without RWLock (or do we have enum of two cases and dispatch / handle differently). Both seem acceptable to me.